text/MichalStobierski/opis.tex

```diff
@@ -0,0 +1,129 @@
+\section{Implementacja}
+Skoro mamy już pomysł na efektywne parsowanie wyjścia, pozostaje jedynie zastanowić się nad implementacją, która sprawi, że znajdzie on zastosowanie w praktyce.\\
+Nieoczywiste jest, jak możemy szybko wyznaczyć prefiks, którego wybór poskutkuje największym progresem przy następnym wypisywaniu wyjścia. Wzorcowy pomysł realizacji $P_{o}=FP$ polega na parsowaniu kolejnych $t_{k}$ dopóki $T[E(i-1)+1:k]$ należy do $D$. Dojdziemy w ten sposób do $T[a:b]=T[E(i-1)+1:L(i-1)]$. W tym momencie wiemy już, że $E(i)$ leży między $a$ i $b$. Chcemy więc wybrać takie $E(i)$, że $L(i)$ będzie możliwie największe. Prosta obserwacja: kandydatami na takie $E(i)$ są tylko te pozycje, dla których $T[E(i)+1:b]$ należy do $D$.\\
+Próbujmy zatem znaleźć kolejnych kandydatów. "Ucinajmy" po literze od przodu $T[a:b]$ (uzyskując $T[a+1:b]$ etc.) dopóki nie trafimy na takie $a'$, że $T[a':b]$ jest w $D$. Wtedy przechodzimy do drugiej fazy, która polega na wydłużaniu $T[a':b]$ (poprzez parsowanie kolejnych $t_{k}$) ponownie dopóki fraza należy do $D$. Uzyskane tak podsłowo $T[a':b']$ ponownie "ucinamy" od przodu i powtarzamy cały proces. Robimy tak dopóki nie wyjdziemy z $a'$ poza $T[a:b]$, czyli dopóki $a' < b$. Nietrudno dowieść, że ostatnie napotkane, prawidłowe $a'$ jest poszukiwaną wartością $E(i)+1$.
+
+No dobrze, ale jak szybko sprawdzać, czy $T[a+x:b]$ jest w $D$ oraz rozszerzać frazy? Drzewa sufiksowe mogą być niezłym pomysłem, użyjemy jednak pary drzew \textit{Trie}: drzewo $T$ będzie utrzymywać po prostu $D$, a $T^{r}$ wszystkie słowa z $D$, ale odwrócone. Ponadto, każda fraza drzewa $D$ będzie posiadać link do odpowiadającej frazy odwrotnej w $T^{r}$ (i w drugą stronę).
+W ten sposób, operacje \textit{insert} i \textit{search} odbywają się jak w normalnym \textit{Trie} (uwaga: \textit{search} pamięta swój ostatni stan, aby po dodaniu jednej litery nie iść znowu od korzenia po całym istniejącym prefiksie), natomiast dodatkowa operacja \textit{contract} bierze węzeł aktualnej frazy, "skacze" po linku do $T^{r}$, wychodzi w górę o jeden węzeł i "wraca" po linku do drzewa $T$. W ten sposób, w czasie $O(1)$ "ucina" pierwszy znak frazy.
+
+Teraz zauważamy, że \textit{contract} wykona się maksymalnie raz na każdy znak tekstu, tak samo \textit{search}, suma operacji \textit{insert} zaś sumarycznie nie przekracza $|T|$, zatem sumarycznie nasz algorytm działa w amortyzowanym czasie $O(|T|)$ - optymalnie!
+
+Jeżeli istotna jest również używana pamięć, to gdy zamiast zwykłych \textit{Trie} użyjemy \textit{skompresowanych Trie} i odpowiednio zadbamy o utrzymywanie etykiet krawędzi w $T^{r}$ to złożoność pamięciowa poprawi się do optymalnego $O(|D|)$.\\
```

Comment:
Is there a technical difficulty that some internal nodes in $T$ are virtual in $T^r$ and vice versa?

Reply:
There is one, because in analysis of $T$ we're using the fact that $D$ has prefix property. Obviously there are $O(|D|)$ nodes in total, and each edge is labeled by just one character, so the total space taken is also $O(|D|)$. 

However, it doesn't apply in reverse case. 

Consider f.e. D = {a, ab, abc}:
$T$ has size 3. But $T^r$ (uncompressed) has size 6. It's not $O(|D|)$, it's rather $O(sum of words lengths)$.
And of course, if you just simply compress $T^r$ you'll get 3 nodes, but labels on the edges are not of length 1 anymore (you have f.e. edge from root labeled by "cba"). 

So to handle this, you can instead of keeping substrings on the edges explicitly use pair of pointers referring to the beginning and ending nodes of the substring in original $T$. It's doable, but has to be implemented very carefully, and I decided not to do it in favor of readability and simplicity of the library.